### PR TITLE
Update Swift tools version to 5.9

### PR DESCRIPTION
**Pull request by d-ronnqvist**
Bug/issue #, if applicable: rdar://116871051

## Summary

This updates SwiftDocC to Swift 5.9 and
- uses shorthand syntax for `if let` (SE-0345)
- modernizes generics syntax (SE-0328, SE-0341, SE-0361)
- removes a no longer needed pack-port of pointer API (SE-0370)
- replaces FileManagerProtocol SPI with package access level (SE-0386)

There are lots more that we _can_ do now, but I feel that those are better addressed in separate PRs. I opted for these changes because they each make the code more readable and have small (but very frequent) and understandable diffs.

## Dependencies

None

## Testing

Nothing in particular. There should be 

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- ~[] Added tests~
- [x] Ran the `./bin/test` script and it succeeded
- ~[ ] Updated documentation if necessary~

### CODE REVIEW

**Comment by d-ronnqvist** [comment]
I made each change as its own commit, to make it easier to review. If you prefer I can break it apart into separate PRs but I didn't feel that that would make it easier to review.

File: Sources/SwiftDocC/Model/Identifier.swift
Comment (patshaughnessy) [inline Sources/SwiftDocC/Model/Identifier.swift:411]: Found a few other places where we could use modern syntax for dictionaries:

DocumentationContext.swift
```
  
    /// The documentation bundles that are currently registered with the context.
    public var registeredBundles: Dictionary<String, DocumentationBundle>.Values {
        return dataProvider.bundles.values
    }
```

BidirectionalMap.swift
```
 func makeIterator() -> Dictionary<Value1, Value2>.Iterator {
        return forward.makeIterator()
    }
```

GroupedSequence.swift
```
    /// Returns an iterator over the members of the sequence.
    func makeIterator() -> Dictionary<Key, Element>.Values.Iterator {
        storage.values.makeIterator()
    }
```

File: Sources/SwiftDocC/Model/Rendering/Diffing/DifferenceBuilder.swift
Comment (patshaughnessy) [inline Sources/SwiftDocC/Model/Rendering/Diffing/DifferenceBuilder.swift:77]: A few other examples of modern syntax for arrays:

MarkupContainer.swift
```
extension MarkupContainer: RandomAccessCollection {
    public subscript(position: Array<Markup>.Index) -> Markup {
        return elements[position]
    }
    
    public var startIndex: Array<Markup>.Index {
        return elements.startIndex
    }
    
    public var endIndex: Array<Markup>.Index {
        return elements.endIndex
    }
}
```

File: Sources/SwiftDocC/Checker/Checker.swift
Comment (patshaughnessy) [inline Sources/SwiftDocC/Checker/Checker.swift:273]: Found a few more places where we could simplify the generic syntax using `some`:

SemanticAnalysis.swift
```
func analyze<Children: Sequence>(
  _ directive: BlockDirective,
  children: Children, source: URL?,
  for bundle: DocumentationBundle,
  in context: DocumentationContext, problems: inout [Problem]
) -> Result where Children.Element == Markup
```

to:
```
func analyze(
   _ directive: BlockDirective,
  children: some Sequence<Markup>,
  source: URL?,
  for bundle: DocumentationBundle,
  in context: DocumentationContext, problems: inout [Problem]
) -> Result
```




DocumentationConverter.swift
```
mutating func convert<OutputConsumer: ConvertOutputConsumer>(
  outputConsumer: OutputConsumer
) throws -> (analysisProblems: [Problem], conversionProblems: [Problem])
```

to:
```
mutating func convert(
  outputConsumer: some ConvertOutputConsumer
) throws -> (analysisProblems: [Problem], conversionProblems: [Problem])
```

DirectiveArgumentWrapper.swift
     
```
func setProperty<T: AutomaticDirectiveConvertible>(
  on containingDirective: T,
  named propertyName: String,
  to any: Any
)
```

to:
```
func setProperty(
  on containingDirective: some AutomaticDirectiveConvertible,
  named propertyName: String,
  to any: Any
)
```


There are many other places in the code where we have generic constraints that appear to be places where we could add `some` - but actually don't compile. The most common example of these are the places where we use the visitor pattern:


Symbol.swift
```
public override func accept<V: SemanticVisitor>(_ visitor: inout V) -> V.Result {
   return visitor.visitSymbol(self)
}
```

Here it appears we could write 
```
public override func accept(_ visitor: inout some SemanticVisitor) -> SemanticVisitor.Result {
   return visitor.visitSymbol(self)
}
```

.... but Swift returns an error: `Cannot access associated type 'Result' from 'SemanticVisitor'; use a concrete type or generic parameter base instead`

Any idea why? Why can't Swift access the `Result` associated type? Is there a better way to write the function declaration in this example?

**Review by patshaughnessy — APPROVED** [review]
Great improvement! Thanks a lot.

I found some additional fixes you can make that follow the same patterns, if you're interested in adding those now. Also various code questions for you.

File: Sources/SwiftDocC/Infrastructure/Diagnostics/DiagnosticConsoleWriter.swift
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Diagnostics/DiagnosticConsoleWriter.swift:39]: I couldn't because it's used in a public protocol and I couldn't use different access levels in the same protocol. 

We will eventually settle on a stable API for it and then we can change the SPI to public API.

File: Sources/SwiftDocC/Indexing/Indexable.swift
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Indexing/Indexable.swift:50]: The first example can't be written as `Collection<DocCSymbolRepresentable>` because that would change its meaning from `where Element: DocCSymbolRepresentable` to `where Element == DocCSymbolRepresentable` (which also doesn't compile). In other places this would be described by `Collection<some DocCSymbolRepresentable>` but syntax wasn't supported in extensions.

File: Sources/SwiftDocC/Model/Identifier.swift
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Model/Identifier.swift:411]: Some of these can be written using opaque types (the `some` keyword) to not define a specific type. For example:
```diff
- public var registeredBundles: Dictionary<String, DocumentationBundle>.Values 
+ public var registeredBundles: some Collection<DocumentationBundle>
```
```diff 
  /// Returns an iterator over the members of the sequence.
- func makeIterator() -> Dictionary<Key, Element>.Values.Iterator {
+ func makeIterator() -> some IteratorProtocol<Element> { 
```

but for the bidirectional map I don't feel like `some IteratorProtocol<(key: Value1, value: Value2)>` is any more readable than `[Value1: Value2].Iterator`

File: Sources/SwiftDocC/Checker/Checker.swift
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Checker/Checker.swift:273]: I didn't realize that I could update the declaration in the protocols themselves. I got some error message about that earlier and thought that it wasn't supported but it looks like it is.

Comment (d-ronnqvist) [reply Sources/SwiftDocC/Checker/Checker.swift:273]: > .... but Swift returns an error: Cannot access associated type 'Result' from 'SemanticVisitor'; use a concrete type or generic parameter base instead

`V.Result` is referring to the specific `Result` associated type of `V`. However `SemanticVisitor.Result` isn't connected to the `some SemanticVisitor` parameter, so it doesn't define a return type that's associated with the parameter type.

File: Sources/SwiftDocC/Model/Rendering/Diffing/DifferenceBuilder.swift
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Model/Rendering/Diffing/DifferenceBuilder.swift:32]: I can't find the specific precedence order but from what I've understood `some A & B` is the same as `some (A & B)`. In other cases I needed to add parentheses around the full expression to write `(some A)?`

**Comment by d-ronnqvist** [comment]
@swift-ci please test

File: Sources/SwiftDocC/Indexing/Indexable.swift
Comment (patshaughnessy) [reply Sources/SwiftDocC/Indexing/Indexable.swift:50]: Got it - I see the new generic syntax refers to the associated type of the parent type.

File: Sources/SwiftDocC/Model/Identifier.swift
Comment (patshaughnessy) [reply Sources/SwiftDocC/Model/Identifier.swift:411]: I agree - `some IteratorProtocol<Element>` is actually quite readable even though it uses the verbose generic syntax. Calling this will return some something that can iterate over Element's.

But yea `some IteratorProtocol<(key: Value1, value: Value2)>` seems confusing either because there's a key and a value in the tuple, or just because of the unhelpful names `Value1` and `Value2`.

File: Sources/SwiftDocC/Checker/Checker.swift
Comment (patshaughnessy) [reply Sources/SwiftDocC/Checker/Checker.swift:273]: > V.Result is referring to the specific Result associated type of V. However SemanticVisitor.Result isn't connected to the some SemanticVisitor parameter, so it doesn't define a return type that's associated with the parameter type.

Ok. But don't all types that conform to `SemanticVisitor` have the same associated type `Result`? It seems to me that Swift should be able to make this compile and run.

File: Sources/SwiftDocC/Indexing/Indexable.swift
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Indexing/Indexable.swift:50]: Specifically the _primary_ associated type. `Collection` is generic over both the `Element` and `Index` but only `Element` is the _primary_ associated type.

**Comment by d-ronnqvist** [comment]
Hm.. the Linux CI seems to be pinned to `SWIFT_VERSION=5.8.1` which causes it to fail.

**Comment by d-ronnqvist** [comment]
This is blocked on the CI (rdar://125077916)

**Comment by d-ronnqvist** [comment]
@swift-ci please test

**Comment by d-ronnqvist** [comment]
The team talked about this today and decided to revert the `package` access level commit and add it back in a follow up PR to unblock this.

**Pull request merged**
